File: 2024/intermediate/14-kotrola-zavorek/solution/solution.c
#include <stdio.h>
#include <stdlib.h>



int evaluate(char end){
    char c = getchar();
    while (c != end){
        if(c == EOF) return 0;

        switch(c){
            case '(': if (!evaluate(')')) return 0; break;
            case '[': if (!evaluate(']')) return 0; break;
            case '{': if (!evaluate('}')) return 0; break;

            case ')':
            case ']':
            case '}': return 0; break;
        }

        c = getchar();
    }
    return 1;
}



int main(){
    
    if(evaluate(EOF)) printf("VALID\n");
    else printf("INVALID\n");

    return 0;
}

File: 2024/intermediate/14-kotrola-zavorek/solution/solution.py
import sys

BRACKETS = {
    '(':')',
    '[':']',
    '{':'}'
}


def getc():
    c = sys.stdin.read(1)
    if c == '': raise EOFError
    return c



def evaluate(end_bracket_type :str):
    try:
        c = getc()
        while True:
            if c in BRACKETS.keys():
                if not evaluate(BRACKETS[c]): return False
                
            if c in BRACKETS.values():
                if end_bracket_type == c: return True
                else: return False
            
            c = getc()

    except EOFError:
        if end_bracket_type == 'EOF':
            return True
        else:
            return False





if __name__ == "__main__":
    if evaluate('EOF'): print("VALID")
    else: print("INVALID")